classes/SpecialMission.php

```diff
@@ -389,32 +391,39 @@ public function nextEvent() {
     /**
      * @throws RuntimeException
      */
-    public function completeMission(): string {
-        // Yen gain for completing the mission
-        $yen_gain = self::$difficulties[$this->difficulty]['yen_per_mission'] * $this->player->rank_num;
-        $yen_gain *= 0.8 + (mt_rand(1, 4) / 10);
-        $yen_gain = floor($yen_gain);
-
-        $this->status = 1;
-        $this->end_time = time();
-        $this->player->addMoney($yen_gain, "Special mission");
-        $this->reward += $yen_gain;
-        $this->player->special_mission = 0;
+    public function completeMission($progress): string {
+        if ($progress > 100) {
+            $progress = 100;
+        }
+        $progress_modifier = $progress / 100;
+        $reward_text = '';
+        if ($progress == 100) {
+            // Yen gain for completing the mission
+            $yen_gain = self::$difficulties[$this->difficulty]['yen_per_mission'] * $this->player->rank_num;
+            $yen_gain *= 0.8 + (mt_rand(1, 4) / 10);
+            $yen_gain = floor($yen_gain);
+
+            $this->status = 1;
+            $this->end_time = time();
+            $this->player->addMoney($yen_gain, "Special mission");
+            $this->reward += $yen_gain;
+            $this->player->special_mission = 0;
 
-        $reward_text = self::$event_names[self::EVENT_COMPLETE_REWARD]['text'] . $yen_gain . '!';
+            $reward_text = self::$event_names[self::EVENT_COMPLETE_REWARD]['text'] . $yen_gain . '!';
 
-        //Reputation Reward
-        if($this->player->reputation->canGain(true)) {
-            $rep_gain = $this->player->reputation->addRep(self::$difficulties[$this->difficulty]['rep_gain']);
-            if($rep_gain > 0) {
-                $this->player->mission_rep_cd = time() + UserReputation::ARENA_MISSION_CD;
-                $reward_text .= ' You have gained ' . $rep_gain . " village reputation!";
+            //Reputation Reward
+            if ($this->player->reputation->canGain(true)) {
+                $rep_gain = $this->player->reputation->addRep(self::$difficulties[$this->difficulty]['rep_gain']);
+                if ($rep_gain > 0) {
+                    $this->player->mission_rep_cd = time() + UserReputation::ARENA_MISSION_CD;
+                    $reward_text .= ' You have gained ' . $rep_gain . " village reputation!";
+                }
             }
         }
 
         $stat_to_gain = $this->player->getTrainingStatForArena();
-        $stat_gain = self::$difficulties[$this->difficulty]['stats_per_mission']
-            + (max(0, $this->player->rank_num - 2) * 2);
+        $stat_gain = (int) (self::$difficulties[$this->difficulty]['stats_per_mission']
+            + (max(0, $this->player->rank_num - 2) * 2)) * $progress_modifier;
```

Comment:
Could be a little more explicit about the intent here:
```suggestion
        $stat_gain = floor(
           (self::$difficulties[$this->difficulty]['stats_per_mission'] + (max(0, $this->player->rank_num - 2) * 2)
           * $progress_modifier
        );
```